xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py

```diff
@@ -0,0 +1,607 @@
+
+_T = TypeVar('_T')
+
+
+class IList(List[_T]):
+    """
+    A list that can be frozen. Once frozen, it can not be modified. 
+    In comparison to FrozenList this supports pattern matching.
+    """
+    _frozen: bool = False
+
+    def freeze(self):
+        self._frozen = True
+
+    def _unfreeze(self):
+        self._frozen = False
+
+    def append(self, __object: _T) -> None:
+        if self._frozen:
+            raise Exception("frozen list can not be modified")
+        return super().append(__object)
+
+    def extend(self, __iterable: Iterable[_T]) -> None:
+        if self._frozen:
+            raise Exception("frozen list can not be modified")
+        return super().extend(__iterable)
+
+    def insert(self, __index: SupportsIndex, __object: _T) -> None:
+        if self._frozen:
+            raise Exception("frozen list can not be modified")
+        return super().insert(__index, __object)
+
+    def remove(self, __value: _T) -> None:
+        if self._frozen:
+            raise Exception("frozen list can not be modified")
+        return super().remove(__value)
+
+    def pop(self, __index: SupportsIndex = ...) -> _T:
+        if self._frozen:
+            raise Exception("frozen list can not be modified")
+        return super().pop(__index)
+
+    def clear(self) -> None:
+        if self._frozen:
+            raise Exception("frozen list can not be modified")
+        return super().clear()
+
+    def __setitem__(self, __index: SupportsIndex, __object: _T) -> None:
+        if self._frozen:
+            raise Exception("frozen list can not be modified")
+        return super().__setitem__(__index, __object)
+
+    def __add__(self, __x: Iterable[_T]) -> IList[_T]:
+        if self._frozen:
+            raise Exception("frozen list can not be modified")
+        if isinstance(__x, IList) and __x._frozen:
+            raise Exception("frozen list can not be modified")
+        return IList(super().__add__(__x))  # type: ignore
+
+    def __iadd__(self, __x: Iterable[_T]) -> IList[_T]:
+        if self._frozen:
+            raise Exception("frozen list can not be modified")
+        return super().__iadd__(__x)  # type: ignore
+
+    def __eq__(self, __o: object) -> bool:
+        if isinstance(__o, IList):
+            return super().__eq__(__o)  # type: ignore
+        return False
+
+
+@dataclass(frozen=True)
+class ISSAValue(ABC):
+    """
+    Represents an immutable SSA variable. An immutable SSA variable is either an operation result
+    or a basic block argument.
+    """
+    typ: Attribute
+    users: IList[IOperation]
+
+    def _add_user(self, op: IOperation):
+        self.users._unfreeze()  # type: ignore
+        self.users.append(op)
+        self.users.freeze()
+
+    def _remove_user(self, op: IOperation):
+        if op not in self.users:
+            raise Exception(
+                f"Trying to remove a user ({op.name}) that is not an actual user of this value!"
```

Comment:
What do you mean with "work in this case"? 

`_remove_user` is internal to the immutable IR and when a user which was not really a user is removed something suspicious is going on so I think an exception for this is valuable.

Reply:
Okay, then let's leave it as it is.
My belief is that `Exception` should be triggered when this is a user error, while bugs in the code should be triggered by `assert`. But since we don't really follow that guideline yet, I'm fine leaving it!